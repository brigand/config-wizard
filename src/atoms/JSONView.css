.JSONView {
  font-size: 16px;
}

.JSONView__Token__string {
  color: red;
}

.JSONView__Token__number {
  color: blue;
}

.JSONView__Token__op {
  color: #666;
}
