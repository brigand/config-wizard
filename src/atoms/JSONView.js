import React, {PropTypes} from 'react';
import './JSONView.css';

export default
class JSONView extends React.Component {
  static propTypes = {
    string: PropTypes.string.isRequired,
  };

  render() {
    const tokens = getTokens(this.props.string);
    return (
      <pre className="JSONView">
        <code>
          {tokens.map((line, i) =>
            <span key={i}>
              {line.map((token, j) =>
                <span
                  key={j}
                  className={`JSONView__Token__${token.type}`}
                >
                  {token.value}
                </span>)
              }
              {'\n'}
            </span>
          )}
        </code>
      </pre>
    );
  }
}

export function getTokens(json) {
  let line = [];
  let buffer = '';
  let isNumber = false;
  let isString = false;

  const lines = [line];

  for (let i = 0; i < json.length; i++) {
    const c = json[i];
    if (c === '\n') {
      line = [];
      lines.push(line);
      buffer = '';
    } else if (isString) {
      if (c === '"') {
        isString = false;
        buffer += c;
        line.push({type: 'string', value: buffer});
        buffer = '';
      } else {
        buffer += c;
      }
    } else if (c === '"') {
      isString = true;
      buffer += c;
    } else if (isNumber) {
      buffer += c;
      if (!/[0-9.]/.test(json[i + 1])) {
        line.push({type: 'number', value: buffer});
        buffer = '';
        isNumber = false;
      }
    } else if (/[0-9.]/.test(c)) {
      isNumber = true;
      buffer += c;
    } else if (['{', '}', '[', ']', ':', ','].indexOf(c) !== -1) {
      line.push({type: 'op', value: c});
      buffer = '';
    } else if (json.slice(i, i + 4) === 'null') {
      line.push({type: 'null'});
      i += 4;
      buffer = '';
    } else if (c === ' ' || c === '\t') {
      buffer += c;
      if (json[i + 1] !== ' ' && json[i + 1] !== '\t') {
        line.push({type: 'whitespace', value: buffer});
        buffer = '';
      }
    }
  }
  return lines;
}
