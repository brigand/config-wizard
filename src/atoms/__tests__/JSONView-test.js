import {getTokens} from '../JSONView';

const example = {
  foo: {
    x: 1,
    y: null,
    z: 'test',
  },
};
const exampleStr = JSON.stringify(example, null, 2);

it(`getTokens doesn't error`, () => {
  const tokens = getTokens(exampleStr);
  expect(Array.isArray(tokens)).toBe(true);
});

it(`getTokens works for simple example`, () => {
  const json = JSON.stringify({x: 123});
  const tokens = getTokens(json);
  console.log(tokens[0])
  expect(tokens).toEqual([
    [
      {type: 'op', value: '{'},
      {type: 'string', value: '"x"'},
      {type: 'op', value: ':'},
      {type: 'number', value: '123'},
      {type: 'op', value: '}'},
    ]
  ]);
});

it(`getTokens works for large example`, () => {
  const tokens = getTokens(exampleStr);
  expect(tokens).toMatchSnapshot();
});
