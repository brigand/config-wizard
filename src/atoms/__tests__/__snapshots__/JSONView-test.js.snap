exports[`test getTokens works for large example 1`] = `
Array [
  Array [
    Object {
      "type": "op",
      "value": "{",
    },
  ],
  Array [
    Object {
      "type": "whitespace",
      "value": "  ",
    },
    Object {
      "type": "string",
      "value": "\"foo\"",
    },
    Object {
      "type": "op",
      "value": ":",
    },
    Object {
      "type": "whitespace",
      "value": " ",
    },
    Object {
      "type": "op",
      "value": "{",
    },
  ],
  Array [
    Object {
      "type": "whitespace",
      "value": "    ",
    },
    Object {
      "type": "string",
      "value": "\"x\"",
    },
    Object {
      "type": "op",
      "value": ":",
    },
    Object {
      "type": "whitespace",
      "value": " ",
    },
    Object {
      "type": "number",
      "value": "1,",
    },
  ],
  Array [
    Object {
      "type": "whitespace",
      "value": "    ",
    },
    Object {
      "type": "string",
      "value": "\"y\"",
    },
    Object {
      "type": "op",
      "value": ":",
    },
    Object {
      "type": "whitespace",
      "value": " ",
    },
    Object {
      "type": "null",
    },
  ],
  Array [
    Object {
      "type": "whitespace",
      "value": "    ",
    },
    Object {
      "type": "string",
      "value": "\"z\"",
    },
    Object {
      "type": "op",
      "value": ":",
    },
    Object {
      "type": "whitespace",
      "value": " ",
    },
    Object {
      "type": "string",
      "value": "\"test\"",
    },
  ],
  Array [
    Object {
      "type": "whitespace",
      "value": "  ",
    },
    Object {
      "type": "op",
      "value": "}",
    },
  ],
  Array [
    Object {
      "type": "op",
      "value": "}",
    },
  ],
]
`;
